### PR TITLE
Add 'color' and 'size' to arguments

### LINKED ISSUE #44670: BUG: Pandas plot does not accept Columns for color keyword
### 

- [ ] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame({
 'A': {1: 6, 2: 2, 3: 1, 4: 8, 5: 6},
 'B': {1: 2, 2: 10, 3: 3, 4: 1, 5: 1},
 'pos': {1: 1, 2: 2, 3: 4, 4: 1, 5: 1},
 'color': {1: 'red', 2: 'blue', 3: 'green', 4: 'red', 5: 'red'}
})

# df >>>
#    A   B  color
# 1  6   2    red
# 2  2  10   blue
# 3  1   3  green
# 4  8   1    red
# 5  6   1    red


df.plot(x='A', y='B', kind='scatter', color=df['color'].values)
# df.plot(x='A', y='B', kind='scatter', color='color')
```

### Issue Description

I tried to plot a scatterplot with different groups, with one color for each group. I thought is is enought to pass the name of the column to the color-keyword, but this throws an error:

`ValueError: 'c' argument must be a color, a sequence of colors, or a sequence of numbers, not ['color']`

### Expected Behavior

I expect this figure, which was generated using the commented line

`s.plot(x='A', y='B', kind='scatter', color=s['color'].values)`

grafik

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.9.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.0-8-amd64
Version          : #1 SMP Debian 5.10.46-4 (2021-08-03)
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.4
numpy            : 1.20.3
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 58.0.4
Cython           : 0.29.24
pytest           : 6.2.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.28.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : 2021.09.0
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.1
sqlalchemy       : 1.4.23
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.54.0
1
​

</details>

### PR DESCRIPTION
**Pull request by mosc9575**
This changes use more intuitive names for 'color' and 'size' and use the same notation like in '_bar_or_line_doc'.

- [ ] closes #44670

### CODE REVIEW

File: pandas/plotting/_core.py
Comment (jreback) [inline pandas/plotting/_core.py:1585]: you can deprecate but not rename as that is a breaking change

these are matpotlib names but not objecting to the change per se

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/plotting/_core.py
Comment (mosc9575) [reply pandas/plotting/_core.py:1585]: I see your point but I need some help to solve it. What is best practice in this situation and how does a deprication warning work?

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1585]: @mosc9575 check some other PRs which deprecate arguments for examples https://github.com/pandas-dev/pandas/pulls?q=is%3Apr+depr+is%3Amerged+

Comment (mosc9575) [reply pandas/plotting/_core.py:1585]: @jreback I used the wrong word. I did not "rename", I added 'size' and `color` to avoid braking code.

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1585]: I wouldn't keep both in the signature, that'd be confusing

I'd suggest:
- use `color` and `size` in the signature
- if `color` is `None`, take `c` from `kwargs` (likewise for `size` and `s`)
- raise an error if someone passes both `color` and `c`
- check that the docs consistently use `color` and `size`

Comment (mosc9575) [reply pandas/plotting/_core.py:1585]: @MarcoGorelli Thank you. I go for your suggestions. I will edit my PR.

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1585]: thanks - will also need tests to cover the new behaviour

File: pandas/plotting/_core.py
Comment (attack68) [inline pandas/plotting/_core.py:1585]: this can be used as positional arguments and your change is not backwards compatible, best to use the order:

`self, x, y, s, c, size, color, **kwargs`

Comment (attack68) [reply pandas/plotting/_core.py:1585]: actually I misread the purpose of this PR, shouldn't we be replicating matplotlibs arguments? Can you indicate why you want to make this change?

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1585]: > this can be used as positional arguments

ah yes, thanks - @mosc9575 please ignore part of my previous comment

let's keep `s` and `c` in the signature then, add `color` and `size` after them, and deprecate `s` and `c`?

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1585]: > Can you indicate why you want to make this change?

Just for consistency really, as in `bar` and `barh` it's `color`

Comment (mosc9575) [reply pandas/plotting/_core.py:1585]: > actually I misread the purpose of this PR, shouldn't we be replicating matplotlibs arguments? Can you indicate why you want to make this change?

I opend this issue, because `c` was not intuitive for me. I don't have the matplotlib background, I normally work with bokeh, and they have the same functionality of `c` named `color`. Then I was searching the pandas documentation and found `color` used for `hbar` and `bar`. Therefor I thought this is not consitent and a misstake by this function. To be honest, I first understood the namen when I looked into matplotlib today. But it is still not "intuitive" to me.

Please let me know if you think this shouldn't changed at all.

Comment (mosc9575) [reply pandas/plotting/_core.py:1585]: I was reading the matplotlib documentation for a whilw and the same problem can be found there. I will wait for a response to my issue [#21795](https://github.com/matplotlib/matplotlib/issues/21795) and change both in one step, if this is approved.

Comment (jreback) [reply pandas/plotting/_core.py:1585]: i think we could also accept color size and and not deprecate

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1585]: Yes, agreed, I'd go with:

- keep signature as is
- if `c` is `None`, take `color` from `kwargs` (likewise for `s` and `size`)
- raise an error if someone passes both `color` and `c`, or `size` and `s` (and test this)

Conversely, for bar:
- keep signature as is
- if `color` is `None`, take `c` from `kwargs`
- raise an error if someone passes both `color` and `c` (and test this)

Comment (mosc9575) [reply pandas/plotting/_core.py:1585]: Matplolib won't change and I think the sugessted changes by @MarcoGorelli are good. I try to edit my PR soon.

**Comment by jreback** [comment]
@mosc9575 if you'd merge master and addrress comments

File: pandas/plotting/_core.py
Comment (MarcoGorelli) [inline pandas/plotting/_core.py:1685]: 🤔  "exactly one" makes it sound like people need to pass exactly one - however, they're allowed to pass neither, right?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Left a comment on error messages (also, will need tests)

File: pandas/plotting/_core.py
Comment (mosc9575) [reply pandas/plotting/_core.py:1685]: This is correct, but I copyied this line from here [https://github.com/pandas-dev/pandas/blob/439906e07dbad5b7047d7e15e37feedd3ae03df8/pandas/tests/plotting/frame/test_frame_color.py#L228] and thought I keep the formulation.

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1685]: ok, that's fine then

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1685]: if you ping when you've added tests I'll take a closer look

Comment (mosc9575) [reply pandas/plotting/_core.py:1691]: Ping. FYI: One Test is not successful, but should be in my opinion. It is the one change in `test_series.py`.

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1691]: can you try moving what you've added to a new test case to check if it still fails there?

Comment (mosc9575) [reply pandas/plotting/_core.py:1691]: It does not fail if I split the test into two. I will update myy PR.

Comment (mosc9575) [reply pandas/plotting/_core.py:1691]: @MarcoGorelli I want to ping one more time, because I think my tasks are done. As far as I see are my added tests successful. To be honest I don't understand the three failures in the pipeline and think that they are not related to my changes. What will happen with this PR in the next few days?

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1691]: yeah they're probably unrelated, don't worry about them

> What will happen with this PR in the next few days?

I'll take a closer look when I get a chance

File: pandas/plotting/_core.py
Comment (attack68) [inline pandas/plotting/_core.py:1134]:
```suggestion
        elif c is not None or color is not None:
```
is safer for future dev

File: pandas/plotting/_core.py
Comment (attack68) [inline pandas/plotting/_core.py:1138]:
```suggestion
        elif c is not None or color is not None:
            kwargs["color"] = color is color is not None else c
```
elif is safer for future dev and no need for setdefault since you have already `pop` ed

File: pandas/plotting/_core.py
Comment (attack68) [inline pandas/plotting/_core.py:1234]:
```suggestion
        elif c is not None or color is not None:
            kwargs["color"] = color if color is not None else c
```

File: pandas/plotting/_core.py
Comment (attack68) [inline pandas/plotting/_core.py:1605]:
```suggestion
    def scatter(self, x, y, s=None, c=None, **kwargs):
```
need to keep the original position arguments to maintain backwards compat for those using positional args.

File: pandas/plotting/_core.py
Comment (attack68) [inline pandas/plotting/_core.py:1688]:
```suggestion
```
`s` is specifically defined

File: pandas/plotting/_core.py
Comment (attack68) [inline pandas/plotting/_core.py:1696]:
```suggestion
        elif s is not None or size is not None:
            kwargs["s"] = s if s is not None else size
```

File: pandas/plotting/_core.py
Comment (attack68) [inline pandas/plotting/_core.py:1698]:
```suggestion
```
`c` is specifically defined

File: pandas/plotting/_core.py
Comment (attack68) [inline pandas/plotting/_core.py:1706]:
```suggestion
        elif c is not None or color is not None:
           kwargs["c"] = c if c is not None else color
```

File: pandas/tests/plotting/test_series.py
Comment (attack68) [inline pandas/tests/plotting/test_series.py:271]:
```suggestion
    @pytest.mark.parametrize("kw", ["c", "color"])
    def test_bar_user_colors(self, kw):
```

File: pandas/tests/plotting/test_series.py
Comment (attack68) [inline pandas/tests/plotting/test_series.py:296]:
```suggestion
        ax = s.plot.bar(**{kw: ["red", "blue", "blue", "red"]})
```
parametrize the test, dont duplicate it

File: pandas/plotting/_core.py
Comment (mosc9575) [reply pandas/plotting/_core.py:1138]: This is a good change. I will remove the typo, it should be ` color if color is not None else c` and I will commit it.

File: pandas/tests/plotting/test_series.py
Comment (mosc9575) [reply pandas/tests/plotting/test_series.py:296]: Thanks a lot for teaching how to parametrize a test.

File: pandas/tests/plotting/frame/test_frame_color.py
Comment (attack68) [inline pandas/tests/plotting/frame/test_frame_color.py:175]:
```suggestion
        ax = df.plot.bar(y="A", **{kw: df["color"]})
        result = [p.get_facecolor() for p in ax.patches]
```

and you can use 

```
@pytest.mark.parametrize("kw", ["color", "c"])
def test_bar_user_colors(self, kw):
````

to avoid code duplication.

**Review by attack68 — CHANGES_REQUESTED** [review]
This is now looking pretty good to me. one or two comments just to remove code duplication in the tests, but it's looking quite minimal now.
After it goes green Im happy to approve. 

Oh, and this should also be added to the doc strings and there should be a release note in the whatsnew if you havent added it already.

**Comment by mosc9575** [comment]
Thanks a lot to all of you for your help and your suggestions. I am learning a lot.

**Comment by mosc9575** [comment]
I have a question. `bar` and  `line` have the same docstring. Now I only changed `bar` and `hbar`, which now accept `c` as a parameter for `color`. This is not implemented for `line` yet. Therefore I didn't change the docstring. Now this looks now odd. Should I also implement this changes to `line` or should I revert the changes for `bar` and `hbar`? In my opinion only excepting `color` is much more intuative and this issue is solved if `scatter` also accept `color`.

**Comment by MarcoGorelli** [comment]
> only accepting color is much more intuative

Is it worth breaking backwards compatibility for this?

I'd suggest either:
- implement this for `line` as well, OR
- de-couple the shared docstring between `bar` and `line`

**Comment by mosc9575** [comment]
If I don't touch `bar` and `hbar` and only add `color` and `size` as keyword arguments  to `scatter` I don't break anything. I am not sure if `bar` and `hbar` really need to accept `c` because this is not used in matplotlib and now it works well with `color`. In this case I don't have to make any changes to the docstring except in the `scatter`function.

**Comment by MarcoGorelli** [comment]
> I am not sure if `bar` and `hbar` really need to accept `c` because this is not used in matplotlib and now it works well with `color`. In this case I don't have to make any changes to the docstring except in the `scatter` function.

Ah I see - yeah sounds good to me

File: pandas/plotting/_core.py
Comment (MarcoGorelli) [inline pandas/plotting/_core.py:1621]: @attack68 thoughts on how this should be documented?

In the original issue ( https://github.com/pandas-dev/pandas/issues/44670#issuecomment-981663172 ) I'd suggested to just silently accept `size` and `color` as aliases without documenting them, not sure they need an extra section in the docs.

Don't know what's best though, I think aliases are quite rare in pandas

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.4.1.rst
Comment (mosc9575) [reply doc/source/whatsnew/v1.4.1.rst:1]: Thanks for the explanation.

File: pandas/plotting/_core.py
Comment (attack68) [reply pandas/plotting/_core.py:1621]: I would maybe just add a one-liner to the already documented `s` arg:

"`size` is also accepted as an alternate keyword argument for `s` for consistency with other methods."

Comment (MarcoGorelli) [reply pandas/plotting/_core.py:1621]: yeah sounds good

Comment (mosc9575) [reply pandas/plotting/_core.py:1621]: I will add this to my changes.

File: doc/source/whatsnew/v1.4.1.rst
Comment (MarcoGorelli) [inline doc/source/whatsnew/v1.4.1.rst:1]: Can you revert changes to this file please?

File: doc/source/whatsnew/v1.5.0.rst
Comment (MarcoGorelli) [inline doc/source/whatsnew/v1.5.0.rst:19]: Lot's of unrelated changes are showing up in this file

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/plotting/_core.py
Comment (MarcoGorelli) [inline pandas/plotting/_core.py:1621]: 1.5.0

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
needs a whatsnew note for 1.5.0

File: pandas/plotting/_core.py
Comment (jreback) [inline pandas/plotting/_core.py:1606]: i think we can list `s, size : str....` here (i don't know if we do this for aliases elsewhere but should follow the same pattern), e.g. see read_csv where we have some aliases

Comment (mosc9575) [reply pandas/plotting/_core.py:1606]: If I follow the syntax of `read_csv` I have to write

```
size : str, scalar or array-like, optional
    Alias for s
```

 I did not find a notation like `s, size: str`.

**Review by MarcoGorelli — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
should we deprecate passing c/s ? (or are these *also* accepted by mpl)?

**Comment by jreback** [comment]
> > are these _also_ accepted by mpl?
> 
> looks like they're inconsistent, `scatter` takes `c`, `bar` takes `color`
> 
> https://matplotlib.org/3.5.0/api/_as_gen/matplotlib.pyplot.scatter.html https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.bar.html

ok over time prob a good idea to use color/size (even if mpl accepts more).

**Comment by jreback** [comment]
thanks @mosc9575

**Comment by mosc9575** [comment]
You are welcome. I appreciated your cooperation and help.

**Pull request merged**
